src/TelemetryChannels/ServerTelemetryChannel/Shared/Implementation/SamplingPercentageEstimatorTelemetryProcessor.cs

```diff
@@ -0,0 +1,244 @@
+        /// </summary>
+        private int currenSamplingRate;
+
+        /// <summary>
+        /// Last date and time sampling percentage was changed.
+        /// </summary>
+        private DateTimeOffset samplingPercentageLastChangeDateTime;
+
+        /// <summary>
+        /// Callback to invoke every time sampling percentage is evaluated.
+        /// </summary>
+        private AdaptiveSamplingPercentageEvaluatedCallback evaluationCallback;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SamplingPercentageEstimatorTelemetryProcessor"/> class.
+        /// <param name="next">Next TelemetryProcessor in call chain.</param>
+        /// </summary>
+        public SamplingPercentageEstimatorTelemetryProcessor(ITelemetryProcessor next)
+            : this(new SamplingPercentageEstimatorSettings(), null, next)
+        {
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="SamplingPercentageEstimatorTelemetryProcessor"/> class.
+        /// <param name="settings">Dynamic sampling estimator settings.</param>
+        /// <param name="callback">Callback to invoke every time sampling percentage is evaluated.</param>
+        /// <param name="next">Next TelemetryProcessor in call chain.</param>
+        /// </summary>
+        public SamplingPercentageEstimatorTelemetryProcessor(
+            SamplingPercentageEstimatorSettings settings, 
+            AdaptiveSamplingPercentageEvaluatedCallback callback, 
+            ITelemetryProcessor next)
+        {
+            if (settings == null)
+            {
+                throw new ArgumentNullException("settings");
+            }
+
+            if (next == null)
+            {
+                throw new ArgumentNullException("next");
+            }
+
+            this.evaluationCallback = callback;
+            this.settings = settings;
+            this.next = next;
+
+            this.currenSamplingRate = settings.EffectiveInitialSamplingRate;
+
+            this.itemCount = new ExponentialMovingAverageCounter(settings.EffectiveMovingAverageRatio);
+
+            this.samplingPercentageLastChangeDateTime = DateTimeOffset.UtcNow;
+
+            // set evaluation interval to default value if it is negative or zero
+            this.evaluationInterval = this.settings.EffectiveEvaluationInterval;
+
+            // set up timer to run math to estimate sampling percentage
+            this.evaluationTimer = new Timer(
+                this.EstimateSamplingPercentage, 
+                null,
+                this.evaluationInterval,
+                this.evaluationInterval);
+        }
+
+        /// <summary>
+        /// Processes telemetry item.
+        /// </summary>
+        /// <param name="item">Telemetry item to process.</param>
+        public void Process(ITelemetry item)
+        {
+            // increment post-samplin telemetry item counter
+            this.itemCount.Increment();
```

Comment:
Originally I had it like that. But then I changed it. My ultimate goal is to reduce telemetry stream overall (including items subject to sampling and not). So, I count all items but sampling percentage I set affects only subject-to-sampling items and that is the way I want it...